gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/FlowGraphPathFinder.java

```diff
@@ -0,0 +1,380 @@
+   * @return a list of {@link FlowEdge}s starting at the srcNode and ending at the destNode.
+   */
+  public Dag<JobExecutionPlan> findPath() throws PathFinderException {
+    try {
+      //Initialization of auxiliary data structures used for path computation
+      this.pathMap = new HashMap<>();
+      this.flowEdgeToSpecExecutorMap = new HashMap<>();
+      this.flowEdgeToMergedConfig = new HashMap<>();
+
+      // Generate flow execution id for this compilation
+      this.flowExecutionId = System.currentTimeMillis();
+
+      //Path computation must be thread-safe to guarantee read consistency. In other words, we prevent concurrent read/write access to the
+      // flow graph.
+      // TODO: we can easily improve the performance by using a ReentrantReadWriteLock associated with the FlowGraph. This will
+      // allow multiple concurrent readers to not be blocked on each other, as long as there are no writers.
+      synchronized (this.flowGraph) {
+        //Base condition 1: Source Node or Dest Node is inactive; return null
+        if (!srcNode.isActive() || !destNode.isActive()) {
+          log.warn("Either source node {} or destination node {} is inactive; skipping path computation.", this.srcNode.getId(),
+              this.destNode.getId());
+          return null;
+        }
+
+        //Base condition 2: Check if we are already at the target. If so, return an empty dag.
+        if ((srcNode.equals(destNode)) && destDatasetDescriptor.contains(srcDatasetDescriptor)) {
+          return new Dag<>(new ArrayList<>());
+        }
+
+        LinkedList<FlowEdgeContext> edgeQueue = new LinkedList<>();
+        edgeQueue.addAll(getNextEdges(srcNode, srcDatasetDescriptor, destDatasetDescriptor, flowConfig));
+        for (FlowEdgeContext flowEdgeContext : edgeQueue) {
+          this.pathMap.put(flowEdgeContext, flowEdgeContext);
+        }
+
+        //At every step, pop an edge E from the edge queue. Mark the edge E as visited. Generate the list of adjacent edges
+        // to the edge E. For each adjacent edge E', do the following:
+        //    1. check if the FlowTemplate described by E' is resolvable using the flowConfig, and
+        //    2. check if the output dataset descriptor of edge E is compatible with the input dataset descriptor of the
+        //       edge E'. If yes, add the edge E' to the edge queue.
+        // If the edge E' satisfies 1 and 2, add it to the edge queue for further consideration.
+        while (!edgeQueue.isEmpty()) {
+          FlowEdgeContext flowEdgeContext = edgeQueue.pop();
+
+          DataNode currentNode = this.flowGraph.getNode(flowEdgeContext.getEdge().getDest());
+          DatasetDescriptor currentOutputDatasetDescriptor = flowEdgeContext.getOutputDatasetDescriptor();
+
+          //Are we done?
+          if (isPathFound(currentNode, destNode, currentOutputDatasetDescriptor, destDatasetDescriptor)) {
+            return buildDag(flowEdgeContext);
```

Comment:
My idea actually meant:
- `FlowGraphPathFinder` doesn't not build the Dag, but return the path
- A new class, possibly named `DagBuilder` builds the Dag from the path. 

In this way, the responsibility is separated so that the way to build the `Dag` can be replaced without touching the logic that focuses on find a path.

Reply:
Introduced a new class FlowGraphPath with method asDag() that returns a Dag from a List<FlowEdgeContext>.